File: src/robot_control/controllers/impedance_controller.py
import numpy as np
import pinocchio as pin
from scipy.spatial.transform import Rotation

from robot_control.robot_wrapper import RobotWrapper
from robot_control.controllers.robot_controller_base import RobotControllerBase


class ImpedanceController(RobotControllerBase):
    def __init__(self, robot, controlled_frame):
        super(ImpedanceController, self).__init__(robot)

        assert isinstance(robot, RobotWrapper)
        self.robot = robot
        self.controlled_frame = controlled_frame

        # Target is assumed to be in a world frame
        self.task_target = pin.SE3(np.eye(3), np.zeros((3, 1)))
        self.task_target_set = False

        # Task space gains
        self.kp = np.diag([10.0, 10.0, 10.0, 10.0, 10.0, 10.0])
        self.kd = np.diag([10, 10, 10, 10, 10, 10])

        # null space variables
        self.kqd_ns = np.diag([0.01] * self.robot.get_dof())
        self.kqp_res = np.diag([0.01] * self.robot.get_dof())
        self.q_rest = self.robot.get_neutral_configuration()

    def set_kp(self, kp):
        self.kp = kp

    def set_kd(self, kd):
        self.kd = kd

    def set_task_target(self, task_target):
        assert isinstance(task_target, pin.SE3)
        self.task_target = task_target
        self.task_target_set = True

    def compute_command(self):
        # Not working yet
        # TODO (giuseppe) switch to end effector dynamics
        if not self.task_target_set:
            print("Task target is not set yet")
            return

        pose_des = self.task_target
        pose_current = self.robot.get_frame_placement(self.controlled_frame)

        # compute rotation error
        dR = Rotation.from_matrix(pose_current.rotation.transpose().dot(pose_des.rotation)).as_quat()
        rotation_err = 2 * dR[3] * dR[:3]
        dp = np.hstack([pose_des.translation - pose_current.translation, rotation_err])

        v = self.robot.get_frame_velocity(self.controlled_frame, pin.ReferenceFrame.LOCAL_WORLD_ALIGNED)
        dv = -v.vector
        f = self.kp.dot(dp) + self.kd.dot(dv)

        self.robot.compute_all_terms()
        j = self.robot.get_frame_jacobian(self.controlled_frame, ref=pin.ReferenceFrame.LOCAL_WORLD_ALIGNED)
        y = j.transpose().dot(f)
        tau = y + self.robot.get_nonlinear_terms()
        return tau

File: robot_control/src/controllers/end_effector_controllers/impedance_controller.cpp
/*!
 * @file     impedance_controller.cpp
 * @author   Giuseppe Rizzi
 * @date     04.12.2020
 * @version  1.0
 * @brief    description
 */

#include "robot_control/controllers/end_effector_controllers/impedance_controller.h"
#include <pinocchio/spatial/explog.hpp>
#include <pinocchio/spatial/se3.hpp>

using namespace Eigen;
namespace pin = pinocchio;

namespace rc {

ImpedanceController::ImpedanceController(std::shared_ptr<RobotWrapper> wrp,
                                         std::string &controlled_frame)
    : controlled_frame(controlled_frame), J(6, wrp->getDof()) {
  wrapper = wrp;
  kp_ = MatrixXd::Identity(6, 6) * 1.0;
  kd_ = MatrixXd::Identity(6, 6) * 1.0;
  target = pin::SE3(Matrix<double, 3, 3>::Identity(), Matrix<double, 3, 1>::Zero());
  std::cout << "Manipulator has " << wrapper->getDof() << " dofs." << std::endl;
  std::cout << "Controlling frame: " << controlled_frame << std::endl;
}

void ImpedanceController::setTaskTarget(pin::SE3 task_target) {
  std::lock_guard<std::mutex> lock(target_mutex);
  target = task_target;
  target_set = true;
}

void ImpedanceController::setKp(const Matrix<double, 6, 1> &kp) { kp_ = kp.asDiagonal(); }

void ImpedanceController::setKd(const Matrix<double, 6, 1> &kd) { kd_ = kd.asDiagonal(); }

VectorXd ImpedanceController::computeCommand() {
  std::lock_guard<std::mutex> lock(target_mutex);
  pin::SE3 &current_pose = wrapper->getFramePlacement(controlled_frame);
  pin::SE3 desired_pose;
  if (target_set) {
    desired_pose = target;
  } else {
    desired_pose = current_pose;
  }

  Matrix<double, 6, 1> error;
  error.head(3) << current_pose.translation() - desired_pose.translation();

  // orientation error
  Quaterniond orientation_d(desired_pose.rotation());
  Quaterniond orientation(current_pose.rotation());
  if (orientation_d.coeffs().dot(orientation.coeffs()) < 0.0) {
    orientation.coeffs() << -orientation.coeffs();
  }
  Eigen::Quaterniond error_quaternion(orientation.inverse() * orientation_d);
  error.tail(3) << error_quaternion.x(), error_quaternion.y(), error_quaternion.z();
  // Transform to base frame
  error.tail(3) << -current_pose.rotation() * error.tail(3);

  Matrix<double, 6, 1> desired_velocity;
  desired_velocity.setZero();

  pin::Motion frame_velocity = wrapper->getFrameVelocity(controlled_frame);
  Matrix<double, 6, 1> current_velocity;
  current_velocity.head<3>() = frame_velocity.linear();
  current_velocity.tail<3>() = frame_velocity.angular();
  Matrix<double, 6, 1> velocity_error = current_velocity - desired_velocity;

  wrapper->getFrameJacobian(controlled_frame, J);
  wrapper->computeAllTerms();

  size_t dof = wrapper->getDof();
  VectorXd tau_task(dof), tau_nullspace(dof), tau_d(dof);

  // Cartesian PD control with damping ratio = 1
  tau_task << J.transpose() * (-kp_ * error - kd_ * velocity_error);

  // kinematic pseuoinverse for nullspace handling
  Eigen::MatrixXd Jt_pinv;
  pseudoInverse(J.transpose(), Jt_pinv);
  VectorXd q = wrapper->getQ();
  VectorXd qd = wrapper->getV();
  tau_nullspace << (MatrixXd::Identity(dof, dof) - J.transpose() * Jt_pinv) *
                       (nullspace_stiffness_ * (q_nullspace_ - q) -
                        (2.0 * sqrt(nullspace_stiffness_)) * qd);

  // Desired torque
  //tau_nullspace.setZero(dof);
  tau_d << tau_task + tau_nullspace + wrapper->getNonLinearTerms();
  return tau_d;
}

VectorXd ImpedanceController::advance(VectorXd &q, VectorXd &v) {
  wrapper->updateState(q, v, true);
  return computeCommand();
}

void ImpedanceController::pseudoInverse(const MatrixXd &m, MatrixXd &m_pinv, bool damped) {
  double lambda_ = damped ? 0.002 : 0.0;

  Eigen::JacobiSVD<Eigen::MatrixXd> svd(m, Eigen::ComputeFullU | Eigen::ComputeFullV);
  Eigen::JacobiSVD<Eigen::MatrixXd>::SingularValuesType sing_vals_ = svd.singularValues();
  Eigen::MatrixXd S_ = m;  // copying the dimensions of m, its content is not needed.
  S_.setZero();

  for (int i = 0; i < sing_vals_.size(); i++)
    S_(i, i) = (sing_vals_(i)) / (sing_vals_(i) * sing_vals_(i) + lambda_ * lambda_);

  m_pinv = Eigen::MatrixXd(svd.matrixV() * S_.transpose() * svd.matrixU().transpose());
}

}  // namespace rc
